packages/utils/src/parser/ParserValidator.ts

```diff
@@ -0,0 +1,123 @@
+import get from 'lodash/get';
+import isEqual from 'lodash/isEqual';
+
+import { ID_KEY } from '../constants';
+import hashForSchema from '../hashForSchema';
+import {
+  CustomValidator,
+  ErrorSchema,
+  ErrorTransformer,
+  FormContextType,
+  RJSFSchema,
+  RJSFValidationError,
+  StrictRJSFSchema,
+  UiSchema,
+  ValidationData,
+  ValidatorType,
+} from '../types';
+
+/** The type of the map of schema hash to schema
+ */
+export type SchemaMap<S extends StrictRJSFSchema = RJSFSchema> = {
+  [hash: string]: S;
+};
+
+/** An implementation of the `ValidatorType` interface that is designed for use in capturing schemas used by the
+ * `isValid()` function. The rest of the implementation of the interface throws errors when it is attempted to be used.
+ * An instance of the object allows the caller to capture the schemas used in calls to the `isValid()` function. These
+ * captured schema, along with the root schema used to construct the object are stored in the map of schemas keyed by
+ * the hashed value of the schema. NOTE: After hashing the schema, an $id with the hash value is added to the
+ * schema IF that schema doesn't already have an $id, prior to putting the schema into the map.
+ */
+export default class ParserValidator<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>
+  implements ValidatorType<T, S, F>
+{
+  /** The rootSchema provided during construction of the class */
+  readonly rootSchema: S;
+
+  /** The map of schemas encountered by the ParserValidator */
+  schemaMap: SchemaMap<S> = {};
+
+  /** Construct the ParserValidator for the given `rootSchema`. This `rootSchema` will be stashed in the `schemaMap`
+   * first.
+   *
+   * @param rootSchema - The root schema against which this validator will be executed
+   */
+  constructor(rootSchema: S) {
+    this.rootSchema = rootSchema;
+    this.addSchema(rootSchema, hashForSchema<S>(rootSchema));
+  }
+
+  /** Adds the given `schema` to the `schemaMap` keyed by the `hash` or `ID_KEY` if present on the `schema`. If the
+   * schema does not have an `ID_KEY`, then the `hash` will be added as the `ID_KEY` to allow the schema to be
+   * associated with it's `hash` for future use (by a schema compiler).
+   *
+   * @param schema - The schema which is to be added to the map
+   * @param hash - The hash value at which to map the schema
+   */
+  addSchema(schema: S, hash: string) {
+    const key = get(schema, ID_KEY, hash);
```

Comment:
Is it possible we may call addSchema with identical schemas? In that case, should each schema be treated as a separate, unique schema?

Reply:
It is possible for this to be called with two identical schemas that, from the JSONSchema perspective are the same, will have the same hash, and need only be treated as a single schema. Replacing the identical schema with itself in the map can easily be avoided with a `if (!has(this.schemaMap, key)) {...}` wrapper around it. At the same time the actual call to set the schema into the map with the key for two identical schemas would, in effect, be a noop.

Reply:
We could use the `btoa` function on the schema then hash that string down to something smaller. I do like the idea of at least preventing same hash different schema issue by throwing an error. I think I'll do that at least.

Reply:
NVM about the `btoa` function, it doesn't make a difference